### DISCUSSION TITLE
Did you consider using Pydantic instead of dataclasses for new mapped column definition?

### DISCUSSION THREAD

**Discussion by XCanG** [sqlalchemy/sqlalchemy#9385] [Ideas]
I still using old method of defining tables, but I wish to migrate to new `mapped_column` syntax, however as soon as I start adding `id` (with autoincrement) to the base class everything stop working because of "Fields without default values cannot appear after fields with default values" and it become a nightmare trying to figure out fix for it. I still didn't manage how to properly write even simplest test project with using this feature and placing `id` in a base class.

I know that dataclasses have `kw_only` parameter, but I couldn't make it to work without modifying sources of sqlalchemy.

However Pydantic on it's root solve this issue and it's overall easier to use than dataclasses.
So maybe SQLAlchemy can use it too for >2.0.0 new mapped classed? I think it will solve a lot of issues and could simplify writing with new syntax.

**Comment by XCanG** [comment]
<details><summary>Also may be someone could help me to sort this out, because I tried various things and none of them works.</summary>

Current version of `base.py`, there a lot of commented code because none of this works for me, but I keep them for reference.
```py
from typing import TypeVar, Type, Any
from sqlalchemy import MetaData, BigInteger, String
from sqlalchemy.ext.declarative import as_declarative, declared_attr
from sqlalchemy.orm import Mapped, DeclarativeBase, MappedAsDataclass, declarative_mixin, registry
from sqlalchemy.sql.base import _NoArg
from pydantic import BaseModel

from .session import metadata as _metadata
from .annotations import int_pk, bigint, str_128, str_128_opt, str_500, str_500_opt

from sqlalchemy import Integer
from sqlalchemy.orm import mapped_column as Column
# from dataclasses import dataclass


BaseModelType = TypeVar("BaseModelType", bound=BaseModel)
models = registry()


# class MappedAsDataclassKW(MappedAsDataclass):
#     def __init__(self) -> None:
#         super().__init__()

#     def __init_sublass__(
#         self,
#         init: _NoArg | bool = _NoArg.NO_ARG,
#         repr: _NoArg | bool = _NoArg.NO_ARG,
#         eq: _NoArg | bool = _NoArg.NO_ARG,
#         order: _NoArg | bool = _NoArg.NO_ARG,
#         unsafe_hash: _NoArg | bool = _NoArg.NO_ARG,
#         match_args: _NoArg | bool = _NoArg.NO_ARG,
#         kw_only: _NoArg | bool = True,
#     ) -> None:
#         return super().__init_subclass__(init, repr, eq, order, unsafe_hash, match_args, kw_only = True)


# @dataclass(kw_only=True)
# class Base(MappedAsDataclass, DeclarativeBase):

# @as_declarative()
# class Base:
#     pass


class MappedBase(DeclarativeBase):
    id: Mapped[int] = Column(Integer, primary_key=True, index=True, autoincrement=True, sort_order=-9999)


# @declarative_mixin
class Base(MappedAsDataclass, MappedBase):
    __abstract__ = True

    metadata: MetaData = _metadata
    id: Mapped[int] = Column(Integer, primary_key=True, index=True, autoincrement=True, init=False, sort_order=-9999, use_existing_column=False)
    # id: Mapped[int_pk]
    # __name__: str

    # type_annotation_map = {
    #     bigint: BigInteger(),
    #     str_128: String(128),
    #     str_128_opt: String(128),
    #     str_500: String(500),
    #     str_500_opt: String(500),
    # }

    # def __init__(self, **kwargs: Any) -> None:
    #     super().__init__()

    #     for key, value in kwargs.items():
    #         if hasattr(self, key):
    #             setattr(self, key, value)

    # def __init_subclass__(
    #     cls,
    #     init: _NoArg | bool = _NoArg.NO_ARG,
    #     repr: _NoArg | bool = _NoArg.NO_ARG,
    #     eq: _NoArg | bool = _NoArg.NO_ARG,
    #     order: _NoArg | bool = _NoArg.NO_ARG,
    #     unsafe_hash: _NoArg | bool = _NoArg.NO_ARG,
    #     match_args: _NoArg | bool = _NoArg.NO_ARG,
    #     kw_only: _NoArg | bool = True,
    # ) -> None:
    #     return super().__init_subclass__(init, repr, eq, order, unsafe_hash, match_args, kw_only = True)

    @declared_attr.directive
    @classmethod
    def __tablename__(cls) -> str:
        return cls.__name__.lower()

    @declared_attr.directive
    @classmethod
    def __mapper_args__(cls) -> dict[str, Any]:
        return {
            "polymorphic_identity": cls.__name__,
            "polymorphic_on": "polymorphic_type",
        }

    @declared_attr
    @classmethod
    def polymorphic_type(cls) -> Mapped[str]:
        return Column(
            String,
            insert_default=cls.__name__,
            init=False,
        )

    def dict(self) -> dict:
        return {key: value for key, value in self.__dict__.items() if not key.startswith("_")}
```

`user.py`, one of example usage, but there other files with similar schemes
```py
from datetime import datetime
from sqlalchemy import Boolean, String, DateTime
from sqlalchemy.orm import Mapped
from sqlalchemy.orm import mapped_column as Column
from sqlalchemy.sql import false, null, func

from .base import Base, models


# @models.mapped_as_dataclass
class User(Base):
    username: Mapped[str] = Column(String, unique=True, index=True, nullable=False)
    email: Mapped[str] = Column(String, unique=True, index=True, nullable=False)
    hashed_password: Mapped[str] = Column(String, nullable=False)

    created_at: Mapped[datetime] = Column(DateTime, server_default=func.now(), nullable=False)
    updated_at: Mapped[datetime | None] = Column(DateTime, server_default=null(), onupdate=datetime.utcnow, nullable=True)

    is_active: Mapped[bool] = Column(Boolean, server_default=true(), nullable=False)
    is_moderator: Mapped[bool] = Column(Boolean, server_default=false(), nullable=False)
    is_superuser: Mapped[bool] = Column(Boolean, server_default=false(), nullable=False)
```
I tried to set `init=False` on `created_at`, `updated_at`, `is_active`, `is_superuser`, but at this moment I removed them because of the problems.

`annotations.py`, I expect to define this file, but commented at this moment because of problems with base class
```py
from typing import Annotated
from sqlalchemy import Integer, BigInteger, String, ForeignKey
from sqlalchemy.orm import mapped_column as Column

# ? Type Aliases

int_pk = Annotated[int, Column(Integer, primary_key=True, index=True, autoincrement=True)]
bigint = Annotated[int, Column(BigInteger, nullable=False)]

# NOTE: Add to DeclarativeBase in "type_annotation_map"
str_128 = Annotated[str, Column(String(128), nullable=False)]
str_128_opt = Annotated[str | None, Column(String(128), default=None, nullable=True)]
str_500 = Annotated[str, Column(String(500), nullable=False)]
str_500_opt = Annotated[str | None, Column(String(500), default=None, nullable=True)]

# ? Foreign Keys

author_fk = Annotated[int, Column(Integer, ForeignKey("author.id", ondelete="CASCADE"), nullable=False)]

user_fk_opt = Annotated[int | None, Column(Integer, ForeignKey("user.id", ondelete="SET NULL"), default=None, nullable=True)]
user_fk_cas = Annotated[int, Column(Integer, ForeignKey("user.id", ondelete="CASCADE"), nullable=False)]
```
</details>

**Reply by CaselIT** [reply]
have you tried removing duplicated id definition? does it work leaving only the one in `Base`?

**Reply by XCanG** [reply]
> have you tried removing duplicated id definition? does it work leaving only the one in `Base`?

It originally didn't have `MappedBase`, it is something I saw today in other discussion that I copy and try to use, original from here: https://github.com/sqlalchemy/sqlalchemy/discussions/9333

**Reply by XCanG** [reply]
From yesterday it look something like that:
```py
class Base(MappedAsDataclass):
    __abstract__ = True

    metadata: MetaData = _metadata
    id: Mapped[int] = Column(Integer, primary_key=True, index=True, autoincrement=True, init=False)
    
    ... other classmethods
```

**Reply by XCanG** [reply]
When I mentioned the way to make it work I edit sources of SQLAlchemy by changing one like where was:
```py
kw_only: Union[_NoArg, bool] = _NoArg.NO_ARG,
```
to `True` value, that was remove error when I created this Base classes, but it's not a good solution, so I keep trying. Also setting it in `__init__` with super didn't solve an issue (didn't affect that), so I throw that idea.

**Reply by CaselIT** [reply]
You can set it for the class using a class argument, as documented https://docs.sqlalchemy.org/en/20/orm/dataclasses.html#class-level-feature-configuration

That's equivalent of using `dataclass(kw_only=True)` to make it apply to all arguments.
Or you can set it per attribute attribute using passing `kw_only=True` on the `mapped_column`

**Reply by XCanG** [reply]
I already tried it. It was in commented code. I may try again on 2.0.4.

Could you tell me, if I make `class User(MappedAsDataclass, Base, repr=False, unsafe_hash=True):` as a base class, do I still need to set repr and unsafe_hash to every table or I could use it only once?

**Reply by XCanG** [reply]
Ok, it seems like only base class will require that, however now I face other issue where it couldn't see User table reference.

**Reply by XCanG** [reply]
Ah, it looks like it now show something like a repr when it try to execute SQL, an example:
```py
[SQL: SELECT "<sqlalchemy.ext.asyncio.engine.AsyncEngine object at 0x000001F2D596C500>"."user".id, "<sqlalchemy.ext.asyncio.engine.AsyncEngine object at 0x000001F2D596C500>"."user".username, "<sqlalchemy.ext.asyncio.engine.AsyncEngine object at 0x000001F2D596C500>"."user".email, "<sqlalchemy.ext.asyncio.engine.AsyncEngine object at 0x000001F2D596C500>"."user".hashed_password, "<sqlalchemy.ext.asyncio.engine.AsyncEngine object at 0x000001F2D596C500>"."user".created_at, "<sqlalchemy.ext.asyncio.engine.AsyncEngine object at 0x000001F2D596C500>"."user".updated_at, "<sqlalchemy.ext.asyncio.engine.AsyncEngine object at 0x000001F2D596C500>"."user".is_moderator, "<sqlalchemy.ext.asyncio.engine.AsyncEngine object at 0x000001F2D596C500>"."user".is_superuser 
FROM "<sqlalchemy.ext.asyncio.engine.AsyncEngine object at 0x000001F2D596C500>"."user" 
WHERE "<sqlalchemy.ext.asyncio.engine.AsyncEngine object at 0x000001F2D596C500>"."user".id = $1::INTEGER]
[parameters: (1,)]
(Background on this error at: https://sqlalche.me/e/20/f405)
```
all columns and tables in "FROM" show as `<sqlalchemy.ext.asyncio.engine.AsyncEngine object at 0x000001F2D596C500>`

**Reply by CaselIT** [reply]
I honestly have no idea how you are getting that in the sql string.

**Comment by CaselIT** [comment]
> I know that dataclasses have kw_only parameter, but I couldn't make it to work without modifying sources of sqlalchemy.

this seems not something that's needed, and you should report an issue if something is not working. edit: seems not https://github.com/sqlalchemy/sqlalchemy/discussions/9385#discussioncomment-5148576

Also note that you don't have to use the default dataclass init, but can use the sqlalchemy one that has no such problem.

> However Pydantic on it's root solve this issue and it's overall easier to use than dataclasses.
> So maybe SQLAlchemy can use it too for >2.0.0 new mapped classed? I think it will solve a lot of issues and could simplify writing with new syntax.

So a few things here:
- dataclasses are builtin in python while pydantic is not, so that's a huge advantage of dataclasses, so in any case it would not be "instead of"
- you don't have to use dataclass, or you can use dataclasses without init, using the default one of sqlalchemy
- pydatinc has a way of working as a dataclass generator, using `pydantic.dataclasses` and you can use it passing the [dataclass_callable](https://docs.sqlalchemy.org/en/20/orm/dataclasses.html#integrating-with-alternate-dataclass-providers-such-as-pydantic). That said I don't know if it changes the init logic of dataclasses
- basing a core functionality on a external library over which sqlalchemy has little control and that could go away at some point is not the best, considering that a somewhat similar alternative is available in the standard library.
- pydantic does validation, in general it does not make much sense validaing the data coming from the db, so it would have a relevant performance impact
- pydantic does not play nice at all with external libraries. In particular SQLAlchemy developers looked into it when writing the initial support for attrs and dataclass for 1.4 and it could not be used. IIRC The part that was causing problems was something on the lines of this one https://github.com/pydantic/pydantic/blob/6267ae3538cf7237ff6769174b4c340415e0f444/pydantic/main.py#LL162-L162C63 that basically throws away anything that's in the class dict and replace it with its own thing.

Also there is [sqlmodel](https://github.com/tiangolo/sqlmodel) that I think has as main thing the integration with pydantic

> Also may be someone could help me to sort this out, because I tried various things and none of them works.

try addinig init=False to the fist id declaration too (or remove it)
there is always the option of not using dataclasses

**Reply by CaselIT** [reply]
just to reiterate that adding support for pydantic is not a no, but
- it would not be instead of dataclasses
- it's not currently a priority

**Reply by XCanG** [reply]
I see, then a couple things: can you improve documentation on site that could have simple example where Base class include `id` key so that for me or others are easy way to start and work with new method and also look at my code in other post and may be point out on issue in it.

**Reply by XCanG** [reply]
Also the reason why I want to use dataclasses is:

1. Because my other part of code already use Pydantic to define things (in FastAPI project), so in a way SQLAlchemy code could be very similar replica to schemes, especially with `Mapped[type]`
2. Because previously there is not good support for type in a table, so when I write the code there was no indication of errors, however with dataclass method in 2.0.0 I can see types and fix potential issues early in IDE.

**Reply by CaselIT** [reply]
Maybe a somewhat complex example could be added to the https://github.com/sqlalchemy/sqlalchemy/tree/main/examples folder and linked from the dataclass section.

if you have an idea of an example that would be useful, feel free to open a PR with it

**Reply by CaselIT** [reply]
> Because previously there is not good support for type in a table, so when I write the code there was no indication of errors, however with dataclass method in 2.0.0 I can see types and fix potential issues early in IDE.

note that other than the init, where you would get `def __init__(self, **kwargs:Any)`, every other type works the same with or without dataclass

**Reply by XCanG** [reply]
To write an example I need to solve my issues first (because it currently not work) and I'm not actually understand differences between `DeclarativeBase`, `MappedAsDataclass`, `@declarative_mixin`, `@as_declarative` and `registry` (decorator). I tried to use them all to make it work, but I only confuse myself and lost in a process.

**Comment by zzzeek** [comment]
Looks like things are being worked out here but overall SQLAlchemy is only able to embrace dataclasses since they are a prominent part of the standard lib.  SQLAlchemy would never have third party high-level libraries as dependencies ,but as mentioned we have a hook that explicitly documents how to use pydantic at https://docs.sqlalchemy.org/en/20/orm/dataclasses.html#integrating-with-alternate-dataclass-providers-such-as-pydantic

**Reply by XCanG** [reply]
Hm, what about the 5th point from CaseIIT post? I thought it wasn't possible, also, the only reason I suggested it that it would solve positional and keyword issue, does this method will do that? If not then I don't see a reason to replace it.

In my other post here with a help from CaseIIT I was somehow able to fix that issue, but another issue was appeared that I still trying to solve (couldn't figure it root and way to fix yet), so with this current system there is a tons of problems. I wish it would be more straightforward and more intuitive.

**Reply by zzzeek** [reply]
if you could refer to the comment and question and show a brief code example of the problem I can comment

**Reply by CaselIT** [reply]
> Hm, what about the 5th point from CaseIIT post? I thought it wasn't possible, also, the only reason I suggested it that it would solve positional and keyword issue, does this method will do that? If not then I don't see a reason to replace it.
> 
> In my other post here with a help from CaseIIT I was somehow able to fix that issue, but another issue was appeared that I still trying to solve (couldn't figure it root and way to fix yet), so with this current system there is a tons of problems. I wish it would be more straightforward and more intuitive.

my comment was regarding normal pydantic classes, extending from BaseModel (or the name of the pydantic base class). it seems that pydantic also has a dataclass like decorator, that behaves more or less as dataclass, and I forgot about it in the reply above, sorry. I will add it as point for future reference.

I don't know if it will change anything regarding init arguments though

**Reply by XCanG** [reply]
> my comment was regarding normal pydantic classes, extending from BaseModel (or the name of the pydantic base class). it seems that pydantic also has a dataclass like decorator, that behaves more or less as dataclass, and I forgot about it in the reply above, sorry. I will add it as point for future reference.

Well, to be specific I didn't need any features from Pydantic except one that allow by default have positional and kwargs without any issues, but other than that is not needed.

**Reply by XCanG** [reply]
> if you could refer to the comment and question and show a brief code example of the problem I can comment

@zzzeek So, if you look at first post it have almost full example for SQLAlchemy setup (of course there is a lot of files from FastAPI, but I can't post every one of them, there is just too much code), but what I did change is for the Base I changed a bit of the way how I initialize it:
```py
class Base(MappedAsDataclass, DeclarativeBase, repr=False, unsafe_hash=True, kw_only=True):
    __abstract__ = True

    metadata: MetaData = _metadata
    id: Mapped[int] = Column(Integer, primary_key=True, index=True, autoincrement=True, init=False, sort_order=-9999, use_existing_column=False)
```

So, before migrating to this I already had CRUD functions to call and they were work before, but now when I call, e.g. get user by id with a code like this:

`app/crud/users.py`
```py
from sqlalchemy.ext.asyncio import AsyncSession
from app.db.models.users import User as DBUser
from app.models.users import User, # ...

async def get_by_id(session: AsyncSession, user_id: int) -> User | None:
    result = await session.execute(select(DBUser).where(DBUser.id == user_id))
    user = result.scalar_one_or_none()
    return User.from_orm(user) if user else None
```
A little note: Since previously I use Pydantic models (which is `User` here), I had names overlap, so I name SQLAlchemy models as `DB__tablename__` (`DBUser` here)

And when it called this code produce SQL text like I show above:
```py
[SQL: SELECT "<sqlalchemy.ext.asyncio.engine.AsyncEngine object at 0x000001F2D596C500>"."user".id, "<sqlalchemy.ext.asyncio.engine.AsyncEngine object at 0x000001F2D596C500>"."user".username, "<sqlalchemy.ext.asyncio.engine.AsyncEngine object at 0x000001F2D596C500>"."user".email, "<sqlalchemy.ext.asyncio.engine.AsyncEngine object at 0x000001F2D596C500>"."user".hashed_password, "<sqlalchemy.ext.asyncio.engine.AsyncEngine object at 0x000001F2D596C500>"."user".created_at, "<sqlalchemy.ext.asyncio.engine.AsyncEngine object at 0x000001F2D596C500>"."user".updated_at, "<sqlalchemy.ext.asyncio.engine.AsyncEngine object at 0x000001F2D596C500>"."user".is_moderator, "<sqlalchemy.ext.asyncio.engine.AsyncEngine object at 0x000001F2D596C500>"."user".is_superuser 
FROM "<sqlalchemy.ext.asyncio.engine.AsyncEngine object at 0x000001F2D596C500>"."user" 
WHERE "<sqlalchemy.ext.asyncio.engine.AsyncEngine object at 0x000001F2D596C500>"."user".id = $1::INTEGER]
[parameters: (1,)]
(Background on this error at: https://sqlalche.me/e/20/f405)
```

I don't know how it become like that, so I'm not sure but I guess it related to Base class somehow.

**Reply by CaselIT** [reply]
I think you need to provide the entire runnable example if you want support with it

**Reply by XCanG** [reply]
If I use `text()` and manually write SQL there is no problem, so only ORM generation is broken here.
e.g. this will work fine:
```py
from sqlalchemy.sql import text

...
    result = await session.execute(text("""SELECT "user".id, "user".username, "user".email, "user".created_at, "user".updated_at FROM "user" WHERE "user".id = 1;"""))
...
```

**Reply by XCanG** [reply]
> I think you need to provide the entire runnable example if you want support with it

I couldn't provide current project, but it is possible to create simplest repo without other stuff to keep it minimal. If you can look at it later, I may create it. _(The only issue would be initializing DB, because as of now I had used `alembic` for automatic migrations and I didn't modify it to setup it on dataclasses yet)_

**Reply by CaselIT** [reply]
I'm only interested in the base model, any mixin and the user table used to generate that sql and how you get that session.

The other part of the example (so fastapi or pydantic or anything else) are not needed.

Again I'm not really sure why you need to use dataclasses here, if you are ever just using the models to query the db, plain classes works perfectly well here, without the need to find the proper init order for dataclasses.

Or even take a look at sqlmodel, that seems to be though from the point of fastapi

**Reply by zzzeek** [reply]
> > if you could refer to the comment and question and show a brief code example of the problem I can comment
> 
> @zzzeek So, if you look at first post it have almost full example for SQLAlchemy setup (of course there is a lot of files from FastAPI, but I can't post every one of them, there is just too much code), but what I did change is for the Base I changed a bit of the way how I initialize it:
> 
> ```python
> class Base(MappedAsDataclass, DeclarativeBase, repr=False, unsafe_hash=True, kw_only=True):
>     __abstract__ = True
> 
>     metadata: MetaData = _metadata
>     id: Mapped[int] = Column(Integer, primary_key=True, index=True, autoincrement=True, init=False, sort_order=-9999, use_existing_column=False)
> ```

Column does not accept those arguments, so I assume you mean mapped_column().

Your issue looks like you are sending the wrong kind of argument to something somewhere.  To get help with issues like these, we communicate the problem using [MCVE](https://stackoverflow.com/help/minimal-reproducible-example) examples.  This allows the problem to be distilled down into its single component and usually by writing one, you'd find the problem in your application just by trying to reproduce it.

In this case an MCVE for the behavior would be as below.  If you can alter it to show how you get your error, it would likely reveal where things went wrong:

```py
from __future__ import annotations

import asyncio
from typing import Any

from sqlalchemy import Integer
from sqlalchemy import MetaData
from sqlalchemy import select
from sqlalchemy.ext.asyncio import AsyncSession
from sqlalchemy.ext.asyncio import create_async_engine
from sqlalchemy.orm import DeclarativeBase
from sqlalchemy.orm import Mapped
from sqlalchemy.orm import mapped_column
from sqlalchemy.orm import MappedAsDataclass

_metadata = MetaData()


class Base(
    MappedAsDataclass,
    DeclarativeBase,
    repr=False,
    unsafe_hash=True,
    kw_only=True,
):
    __abstract__ = True

    metadata: MetaData = _metadata
    id: Mapped[int] = mapped_column(
        Integer,
        primary_key=True,
        index=True,
        autoincrement=True,
        init=False,
        sort_order=-9999,
        use_existing_column=False,
    )


class User(Base):
    __tablename__ = "a"

    data: Mapped[str]


async def get_by_id(session: AsyncSession, user_id: int) -> Any:
    result = await session.execute(select(User).where(User.id == user_id))
    user = result.scalar_one_or_none()
    return user


async def main():
    e = create_async_engine("sqlite+aiosqlite://")

    async with e.begin() as conn:
        await conn.run_sync(Base.metadata.create_all)

    async with AsyncSession(e) as session:

        u1 = User(data="u5")
        session.add(u1)
        await session.flush()
        user_id = u1.id

        print(await get_by_id(session, user_id))


asyncio.run(main())
```

**Reply by XCanG** [reply]
> Column does not accept those arguments, so I assume you mean mapped_column().

Yes, when I imported them I imported them like `from sqlalchemy.orm import mapped_column as Column`

**Reply by XCanG** [reply]
I'm not sure if I could make it into MCVE, but I making minimal copy of a project right now and removing unnecessary stuff.

**Reply by zzzeek** [reply]
you don't even have to, just run the program I have above and try to compare it to your application and you hopefully can find what's wrong independently.   it looks like you are passing an Engine object somewhere, like `__table_args__ = {"schema": my_engine}` or something weird like that.   This is not any foundational problem with dataclasses it's just you have a argument error somewhere.

**Reply by XCanG** [reply]
@zzzeek or @CaselIT I make a simpler copy of a project, it not MCVE, but whole project with almost all of it being removed.
I only keep `User` model and all functions related to it.

Endpoints that don't work: get user by id, get user by username, list users, create user
Endpoints that work (`text` is used): list users with filters, update user, delete user

I archive code and upload on GDrive: https://drive.google.com/file/d/1wjuyz5OCg673YgXF7Uz6y80VVcNhewwf/view?usp=share_link

Main file for SQLAlchemy in `/app/db/base.py` and in `models` subfolder `users.py`

**Reply by XCanG** [reply]
> you don't even have to, just run the program I have above and try to compare it to your application and you hopefully can find what's wrong independently. it looks like you are passing an Engine object somewhere, like `__table_args__ = {"schema": my_engine}` or something weird like that. This is not any foundational problem with dataclasses it's just you have a argument error somewhere.

Result:
```sh
> python .\test.py
<__main__.User object at 0x00000207B140E510>
```

While it work, I couldn't see real difference, because it looks like I use same base, but something else is broken...

**Reply by CaselIT** [reply]
> @zzzeek or @CaselIT I make a simpler copy of a project, it not MCVE, but whole project with almost all of it being removed. I only keep `User` model and all functions related to it.
> 
> Endpoints that don't work: get user by id, get user by username, list users, create user Endpoints that work (`text` is used): list users with filters, update user, delete user
> 
> I archive code and upload on GDrive: https://drive.google.com/file/d/1wjuyz5OCg673YgXF7Uz6y80VVcNhewwf/view?usp=share_link
> 
> Main file for SQLAlchemy in `/app/db/base.py` and in `models` subfolder `users.py`

taking a look. first think first, you should look into sql injection https://en.wikipedia.org/wiki/SQL_injection 
how you are doing string queries is **not** how you should do, please use bound parameters passed to `execute()`. see here https://docs.sqlalchemy.org/en/20/core/sqlelement.html#sqlalchemy.sql.expression.text

**Reply by CaselIT** [reply]
found the issue, you are passing engine as first element of MetaData. that doesn't work and it gets used as schema

I'm surprised it doesn't raise

**Reply by XCanG** [reply]
> found the issue, you are passing engine as first element of MetaData. that doesn't work and it gets used as schema
> 
> I'm surprised it doesn't raise

What the correct way then?

**Reply by CaselIT** [reply]
remove engine from the `MetaData` constructor

**Reply by XCanG** [reply]
> taking a look. first think first, you should look into sql injection https://en.wikipedia.org/wiki/SQL_injection how you are doing string queries is **not** how you should do, please use bound parameters passed to `execute()`. see here https://docs.sqlalchemy.org/en/20/core/sqlelement.html#sqlalchemy.sql.expression.text

I did test for injections and couldn't make them work. Firstly major stuff could be not passed by Pydantic checks, then I also escape string ends in `_SQL()` function, so it couldn't be closed within text content.

**Reply by XCanG** [reply]
> remove engine from the `MetaData` constructor

Thanks, it work now. I don't remember why it was added here, but I do remember that at some time I saw type error here where it was sync engine and not async, so I add `# type: ignore` here.

**Reply by CaselIT** [reply]
> > taking a look. first think first, you should look into sql injection https://en.wikipedia.org/wiki/SQL_injection how you are doing string queries is **not** how you should do, please use bound parameters passed to `execute()`. see here https://docs.sqlalchemy.org/en/20/core/sqlelement.html#sqlalchemy.sql.expression.text
> 
> I did test for injections and couldn't make them work. Firstly major stuff could be not passed by Pydantic checks, then I also escape string ends in `_SQL()` function, so it couldn't be closed within text content.

you are really making things very complex for yourself there, when the standard things is to just let the db or driver to this for you.
